flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/nn/QuadTree.scala

```diff
@@ -0,0 +1,340 @@
+     * @param message
+     */
+    case class metricException(message: String) extends Exception(message)
+
+    /**
+     * minDist is defined so that every point in the box
+     * has distance to queryPoint greater than minDist
+     * (minDist adopted from "Nearest Neighbors Queries" by N. Roussopoulos et al.)
+     *
+     * @param queryPoint
+     * @return
+     */
+
+    def minDist(queryPoint: Vector): Double = {
+      var minDist = 0.0
+      for (i <- 0 to queryPoint.size - 1) {
+        if (queryPoint(i) < center(i) - width(i) / 2) {
+          minDist += math.pow(queryPoint(i) - center(i) + width(i) / 2, 2)
+        } else if (queryPoint(i) > center(i) + width(i) / 2) {
+          minDist += math.pow(queryPoint(i) - center(i) - width(i) / 2, 2)
+        }
+      }
+
+      if (distMetric.isInstanceOf[SquaredEuclideanDistanceMetric]) {
+        minDist
+      } else if (distMetric.isInstanceOf[EuclideanDistanceMetric]) {
+        math.sqrt(minDist)
+      } else{
+        throw metricException(s" Error: metric must be Euclidean or SquaredEuclidean!")
+      }
+    }
+
+    /**
+     * Finds which child queryPoint lies in.  node.children is a Seq[Node], and
+     * whichChild finds the appropriate index of that Seq.
+     * @param queryPoint
+     * @return
+     */
+    def whichChild(queryPoint: Vector): Int = {
+      var count = 0
+      for (i <- 0 to queryPoint.size - 1) {
+        if (queryPoint(i) > center(i)) {
+          count += Math.pow(2, queryPoint.size -1 - i).toInt
+        }
+      }
+      count
+    }
+
+    def makeChildren() {
+      val centerClone = center.copy
+      val cPart = partitionBox(centerClone, width)
+      val mappedWidth = 0.5*width.asBreeze
+      children = cPart.map(p => new Node(p, mappedWidth.fromBreeze, null))
+
+    }
+
+    /**
+     * Recursive function that partitions a n-dim box by taking the (n-1) dimensional
+     * plane through the center of the box keeping the n-th coordinate fixed,
+     * then shifting it in the n-th direction up and down
+     * and recursively applying partitionBox to the two shifted (n-1) dimensional planes.
+     *
+     * @param center the center of the box
+     * @param width a vector of lengths of each dimension of the box
+     * @return
+     */
+    def partitionBox(center: Vector, width: Vector): Seq[Vector] = {
+
+      def partitionHelper(box: Seq[Vector], dim: Int): Seq[Vector] = {
+        if (dim >= width.size) {
+          box
+        } else {
+          val newBox = box.flatMap {
+            vector =>
+              val (up, down) = (vector.copy, vector)
+              up.update(dim, up(dim) - width(dim) / 4)
+              down.update(dim, down(dim) + width(dim) / 4)
+
+              Seq(up,down)
+          }
+          partitionHelper(newBox, dim + 1)
+        }
+      }
+      partitionHelper(Seq(center), 0)
+    }
+  }
+
+
+  val root = new Node( ((minVec.asBreeze + maxVec.asBreeze)*0.5).fromBreeze,
```

Comment:
Please reformat this line. Remove a space after parenthesis and add spaces around `*`.